### DISCUSSION TITLE
Raw format

### DISCUSSION THREAD

**Discussion by lboclboc** [lboclboc/indi-3rdparty#9] [General]
> QRainman commented 6 hours ago
> One more thing: I am not sure if the 2 LSBs for every pixel are set correctly. The only test i can think of right now would be to put > an appropriate color filter in front of the camera and check which bits go up. However the pedestal is already > 2 bit and the pixel > noise would make this almost impossible.

I am not sure either. I have only done a subjective judgment. A friend of mine suggested to shine a red laser and just precisely overexpose the red channel, thus causing all red LSB:s to hard 1:s, and perhaps green and blue lsb:s to vary. Could work.

**Comment by QRainman** [comment]
Good idea. I'll give that a try.

**Comment by QRainman** [comment]
Ok, not so easy if all you have is a bunch of broken binoculars and a laser pointer, but i think i managed to get a few pixels:
```
FF AB FF 83 B7
FF 75 FF B1 BF
FB 6E FF 6A 3F
FF 93 FF 82 B7
FC 54 FF 56 FB
FF FB FF CE FB
F7 6C FF C4 3C
F0 67 FF 90 F6
```
Especially with that last line it looks like what you had originally is the correct order (this is on V1):
```
                cur_row[x-1] = (cur_row[x-1] | ((byte >> 6) & 0x03)) << (16-10); // Merge bits together and upshift from bit9 to bit15.
                cur_row[x-2] = (cur_row[x-2] | ((byte >> 4) & 0x03)) << (16-10);
                cur_row[x-3] = (cur_row[x-3] | ((byte >> 2) & 0x03)) << (16-10);
                cur_row[x-4] = (cur_row[x-4] | ((byte >> 0) & 0x03)) << (16-10);
```

Now i have to figure out how to twiddle those bits in the other part of the code...

**Comment by QRainman** [comment]
Alright, the original u32Magic pattern was correct. I amended the PR.

**Comment by lboclboc** [comment]
Nice work. Do you consider it right now? I can test on my cameras also now in that case.

**Reply by QRainman** [reply]
Yes, should be correct now *fingers crossed :)

**Comment by lboclboc** [comment]
For commit 23cd80b, V2 looks perfect. The V1 image, with auto-debayer in kstars looks wrong with purble stripes. Looks like its RGGB instead of BGGR when I check a non-debayered image. I think the sensor should produce a BGGR but I am not sure. Ill take a raw picture with raspistill and check in there, it should be ovious..

**Comment by lboclboc** [comment]
This is weird, this is a zoomed up snip from first row of the raw jpg from raspistill, seems like its rather bbggm  (m = low mixed bits). 
![v1-raw-snip](https://user-images.githubusercontent.com/10403183/103928642-6e83d600-511c-11eb-9514-b1ca0ec610df.png)

**Reply by QRainman** [reply]
do you have hflip or vflip on by any chance? Also, is this V1 or V2?

**Reply by lboclboc** [reply]
It's v1, and just made with: raspistill -r -n - t 1 -o raw.jpg
Then the jpg and brcm part is removed

**Reply by QRainman** [reply]
I really don't understand this. That hex dump up there is also from my V1 with the laser hitting it doing and it shows the alternating pattern between Red and Green. raspistill command was:
```
raspistill -r -ex off -o red-laser-raw-0.25s.jpg -th none -awb off -ss 500000  -awbg 1.0,1.0 -ag 1.0 -dg 1.0 -v -set
```

The picture below is from Kstars with V1 camera for https://github.com/lboclboc/indi-3rdparty/commit/23cd80ba4fdd4d58f6ccf3b30db7d915ae04e738

![V1-GreenRedBlueMatching-1s-G7 0-small](https://user-images.githubusercontent.com/18513888/103948977-f678d880-513a-11eb-9389-748006a2686a.png)

I did purge the config in ~/.indi/Raspi Cam Something.xml before taking the picture to be sure the bayer pattern config was not overwritten by the config file though.

**Comment by lboclboc** [comment]
Sorry, i must have done something wrong yesterday. I used your exact settings to take a new picture and this is what I got:

![red-laser-raw-0 25s](https://user-images.githubusercontent.com/10403183/104002101-e09dfe80-51a0-11eb-8729-cb901a25aa50.png)

Looks more like GBRG coded as G,B,G,B,mix here. When I re-check my test from yesterday but correctly, then that also looks like GBRG
and that is what you have used in your code also. Should work then...

**Reply by lboclboc** [reply]
Ok, so after cleaning the .indi/Rpi... on the raspberry it works fine... Thanks for your PR. Ill submit it shortly

**Reply by QRainman** [reply]
Agree. This is the top left corner of an image i took earlier of some red object:
![V1-RedBag-0 5s-raw-TopLeft](https://user-images.githubusercontent.com/18513888/104009190-e862a080-51aa-11eb-8980-88a5120be240.png)
It's unfortunately not perfect, but the red is about 1.35 thimes the blue and green components, which are approximately the same.
Here's the hexdump from the raw file:
Start of first data line:
```
2E 2F 2F 2D 34   
2E 30 35 2F DA 
37 2F 33 2E D2
```
Start of second data line:
```
51 34 59 35 41 
58 37 55 37 1D 
47 33 46 2F 97
```
Sure looks like the second line does have the red in it and starts with a red pixel. So, yeah, GBRG it is and as so often: no matter how detailed the documentation seems like, if you don't check everything yourself you are screwed...
They probably flip the image in the firmware during some processing step. Anyway, i think this clears up the rest my doubts for our decoder for me. V1 is GBRG and V2 is RGGB.

**Reply by QRainman** [reply]
> 
> 
> Ok, so after cleaning the .indi/Rpi... on the raspberry it works fine... Thanks for your PR. Ill submit it shortly

For that, maybe we can read a serial number or the chip type and make it part of the device name, so when it gets saved, the settings  depend on the type of camera so there are no mixups. Otoh, i don't know how many V1s are still out there in service and if that wouldn't be overkill.

**Comment by lboclboc** [comment]
I don't think its a real problem since its on the rpi it self, and there are probably very few people using multiple cameras on the RPI. Sure in upgrade situations but then cleaning the cache is good enough I think.
Ive just created a pr now to Jasem. Ive tested your changes on all cameras and its working fine (I did one minor change first).

**Reply by QRainman** [reply]
Alright, time to look into binning now :)

**Comment by lboclboc** [comment]
Exactly my idea also. Not sure how much time ill have though on this for a while.

**Comment by QRainman** [comment]
Would binning just combine neighboring pixels or rather pixels with equal colors?

**Reply by lboclboc** [reply]
The Hiq has hw-support for binning but I have not started to look into that. I would assume with equal colors, otherwise its not very useful. Not sure about the older cameras if they have hw binning.

**Reply by QRainman** [reply]
They kind of do by setting the right resolution, but i think this is ultimately done by either the firmware or the driver, not on the chip itself. If i understood it right, then the raw output will always be un-binned and we would have to do this ourselves in the decoding stage.

**Reply by lboclboc** [reply]
Ok, sorry I have very little knowledge about how binning is actually made in a camera. I thought binning was made on the silicon level to lower the noise since that would be before the AD converters. I'll try to bring in a friend here that has a lot of knowledge of sensors..

**Reply by QRainman** [reply]
I'm also not sure how it works tbh. Table 16 on page 53  [in the manual for V2](https://raw.githubusercontent.com/circuitvalley/mipi_csi_receiver_FPGA/master/Camera%20Datasheet/RASPBERRY%20PI%20CAMERA%20V2%20DATASHEET%20IMX219PQH5_7.0.0_Datasheet_XXX.PDF) suggests that x2 binning is actually done in analog mode. Given my (admittedly very rudimentary) knowledge of CCDs i don't understand how that would work though. To get same color pixels, they would have to essentially skip over pixels when reading out and averaging.

**Reply by QRainman** [reply]
These guys [here](https://picamera.readthedocs.io/en/release-1.13/fov.html#sensor-modes) suggest that binning only works in video capture mode, which i suppose one  could do, but i don't know if it can spit out raw frames in that mode.

**Reply by QRainman** [reply]
Aha!
```
raspistill -r -ex off -o binning.jpg -th none -awb off -ss 500000  -awbg 1.0,1.0 -ag 1.0 -dg 1.0 -v -set -w 1296 -h 972 -md 4
```
```
BRCMo...ü.......ov5647 version 0.1...............................................................................................................¡......ÿ.......`...............2x2_BINNED
```
I think we might be on to something here!
Raw block length is 1592832 bytes which is 1296x982.22, which - with padding - looks like the right size for a 2x2 binned V1 camera =)

**Reply by QRainman** [reply]
As a collateral,  i think setting the sensor mode explicitly could also solve our exposure time problems, because the damn thing would then have no choice but to select the correct frame rate mode.

**Reply by lboclboc** [reply]
haha, yes. Ive just used auto-mode because of not knowing better. Specifically setting it could very well make the frame-rate deterministic yes. Good point.

**Reply by lboclboc** [reply]
When I get some time over Ill try to force the mode for the v1 and v2 cameras but I am still looking for how to find the correct modes for the HiQ camera.

**Reply by QRainman** [reply]
You saw this discussion here?
https://www.raspberrypi.org/forums/viewtopic.php?t=276856
Or is it more a question of how to set the mode correctly?

**Comment by lboclboc** [comment]
Thanks, I did miss that. It also contained an updated link to the camera documentation I have not seen before where the HiQ camera is included. Should be quite what obvious what modes to use now then.
